File: packages/go/style.go
package rasterstyle

import (
	"encoding/json"
	"strings"
	"sync"

	"github.com/mapseekai/raster-style/packages/go/internal/assets"
	"github.com/santhosh-tekuri/jsonschema/v6"
)

// Style is a validated, normalized document. Its fields are intentionally private:
// preserving omitted fields and union variants is part of the wire contract.
// A successfully constructed Style is immutable and safe for concurrent readers.
type Style struct {
	document  map[string]any
	canonical []byte
}

var schemaOnce sync.Once
var schemaValidator *jsonschema.Schema
var schemaError error

func loadSchema() (*jsonschema.Schema, error) {
	schemaOnce.Do(func() {
		var document any
		if schemaError = json.Unmarshal(assets.Schema, &document); schemaError != nil {
			return
		}
		compiler := jsonschema.NewCompiler()
		schemaError = compiler.AddResource("urn:mapseek:raster-style:2.0", document)
		if schemaError != nil {
			return
		}
		schemaValidator, schemaError = compiler.Compile("urn:mapseek:raster-style:2.0")
	})
	if schemaError != nil {
		return nil, failure("E_CONFIG", "Embedded schema could not be compiled", "")
	}
	return schemaValidator, nil
}

func newStyle(value any) (*Style, error) {
	validator, err := loadSchema()
	if err != nil {
		return nil, err
	}
	if err := validator.Validate(value); err != nil {
		return nil, failure("E_SCHEMA", "Raster style does not match draft.2 schema", "")
	}
	document := value.(map[string]any)
	if err := validateSemantics(document); err != nil {
		return nil, err
	}
	normalizeColors(document)
	canonical, err := canonicalJSON(document)
	if err != nil {
		return nil, err
	}
	return &Style{document: document, canonical: canonical}, nil
}

// ParseJSON validates the complete JSON style; it never inserts renderer defaults.
func ParseJSON(data []byte) (*Style, error) {
	value, err := parseJSONStrict(data)
	if err != nil {
		return nil, err
	}
	return newStyle(value)
}

// MarshalJSON enables json.Marshal(style) and Fiber's c.JSON(style).
func (style Style) MarshalJSON() ([]byte, error) {
	if style.document == nil {
		return nil, failure("E_SCHEMA", "Uninitialized Style", "")
	}
	return append([]byte(nil), style.canonical...), nil
}

// UnmarshalJSON validates before replacing the receiver; failures leave it intact.
func (style *Style) UnmarshalJSON(data []byte) error {
	parsed, err := ParseJSON(data)
	if err != nil {
		return err
	}
	*style = *parsed
	return nil
}

// JSON returns an owned copy; callers cannot mutate the validated document.
func (style Style) JSON() []byte { return append([]byte(nil), style.canonical...) }

// Document returns an independent map for callers that need to edit a style.
// Submit edits through ParseJSON before encoding again.
func (style Style) Document() map[string]any {
	return cloneObject(style.document)
}

func cloneObject(source map[string]any) map[string]any {
	if source == nil {
		return nil
	}
	result := make(map[string]any, len(source))
	for key, value := range source {
		result[key] = cloneValue(value)
	}
	return result
}
func cloneValue(value any) any {
	switch current := value.(type) {
	case map[string]any:
		return cloneObject(current)
	case []any:
		result := make([]any, len(current))
		for index, item := range current {
			result[index] = cloneValue(item)
		}
		return result
	default:
		return value
	}
}

func objectAt(object map[string]any, key string) map[string]any {
	result, _ := object[key].(map[string]any)
	return result
}
func arrayAt(object map[string]any, key string) []any {
	result, _ := object[key].([]any)
	return result
}
func getPath(object map[string]any, path string) (any, bool) {
	parts := strings.Split(path, ".")
	for _, key := range parts[:len(parts)-1] {
		object = objectAt(object, key)
		if object == nil {
			return nil, false
		}
	}
	value, present := object[parts[len(parts)-1]]
	return value, present
}
func setPath(object map[string]any, path string, value any) {
	parts := strings.Split(path, ".")
	for _, key := range parts[:len(parts)-1] {
		next := objectAt(object, key)
		if next == nil {
			next = make(map[string]any)
			object[key] = next
		}
		object = next
	}
	object[parts[len(parts)-1]] = value
}
func normalizeColor(color string) string {
	color = strings.ToLower(color)
	if len(color) == 7 {
		color += "ff"
	}
	return color
}
func normalizeColors(style map[string]any) {
	for _, path := range []string{"renderer.color", "opacity.nodata_color", "image.background"} {
		if color, found := getPath(style, path); found {
			setPath(style, path, normalizeColor(color.(string)))
		}
	}
	colorMap := objectAt(objectAt(style, "renderer"), "color_map")
	for _, key := range []string{"under", "over", "outside_color", "fallback_color"} {
		if color, ok := colorMap[key].(string); ok && strings.HasPrefix(color, "#") {
			colorMap[key] = normalizeColor(color)
		}
	}
	for _, key := range []string{"stops", "entries"} {
		for _, item := range arrayAt(colorMap, key) {
			entry := item.(map[string]any)
			entry["color"] = normalizeColor(entry["color"].(string))
		}
	}
	for index, color := range arrayAt(colorMap, "colors") {
		colorMap["colors"].([]any)[index] = normalizeColor(color.(string))
	}
}

File: crates/raster-style/src/style.rs
use std::cell::Cell;
use std::sync::OnceLock;

use serde::{de, de::DeserializeSeed, Deserialize, Deserializer, Serialize};
use serde_json::Value;

use crate::error::{Error, Result};
use crate::json::{canonical_value, parse_strict, StrictSeed};
use crate::semantics::validate_semantics;

const SCHEMA: &str = include_str!("../assets/raster-style-v2.schema.json");
static VALIDATOR: OnceLock<std::result::Result<jsonschema::Validator, String>> = OnceLock::new();

pub(crate) fn validator() -> Result<&'static jsonschema::Validator> {
    VALIDATOR
        .get_or_init(|| {
            let schema: Value = serde_json::from_str(SCHEMA).map_err(|error| error.to_string())?;
            jsonschema::validator_for(&schema).map_err(|error| error.to_string())
        })
        .as_ref()
        .map_err(|_| Error::new("E_CONFIG", "Embedded schema could not be compiled"))
}

/// A validated, normalized document. No mutable reference to its value is exposed.
#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(transparent)]
pub struct Style(Value);

impl Style {
    pub fn from_json(text: &str) -> Result<Self> {
        Self::from_validated_value(parse_strict(text)?)
    }

    /// Programmatic values take the same binary64, budget and validation path as raw JSON.
    pub fn from_value(value: Value) -> Result<Self> {
        Self::from_json(&canonical_value(&value)?)
    }

    pub(crate) fn from_validated_value(mut value: Value) -> Result<Self> {
        if !validator()?.is_valid(&value) {
            return Err(Error::new(
                "E_SCHEMA",
                "Raster style does not match draft.2 schema",
            ));
        }
        validate_semantics(&value)?;
        normalize_colors(&mut value);
        canonical_value(&value)?;
        Ok(Self(value))
    }

    pub fn as_value(&self) -> &Value {
        &self.0
    }
    pub fn into_value(self) -> Value {
        self.0
    }
    pub fn to_json(&self) -> Result<String> {
        canonical_value(&self.0)
    }
}

impl<'de> Deserialize<'de> for Style {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> std::result::Result<Self, D::Error> {
        let nodes = Cell::new(0);
        let value = StrictSeed::root(&nodes).deserialize(deserializer)?;
        Self::from_validated_value(value).map_err(de::Error::custom)
    }
}

fn normalize_color(value: &mut Value) {
    if let Some(color) = value.as_str() {
        let mut normalized = color.to_ascii_lowercase();
        if normalized.len() == 7 {
            normalized.push_str("ff");
        }
        *value = Value::String(normalized);
    }
}

fn normalize_colors(style: &mut Value) {
    for path in [
        "/renderer/color",
        "/opacity/nodata_color",
        "/output/background",
    ] {
        if let Some(color) = style.pointer_mut(path) {
            normalize_color(color);
        }
    }
    let Some(color_map) = style.pointer_mut("/renderer/color_map") else {
        return;
    };
    for key in ["under", "over", "outside_color", "fallback_color"] {
        if let Some(color) = color_map.get_mut(key) {
            if color.as_str().is_some_and(|color| color.starts_with('#')) {
                normalize_color(color);
            }
        }
    }
    for key in ["stops", "entries"] {
        if let Some(entries) = color_map.get_mut(key).and_then(Value::as_array_mut) {
            for entry in entries {
                normalize_color(&mut entry["color"]);
            }
        }
    }
    if let Some(colors) = color_map.get_mut("colors").and_then(Value::as_array_mut) {
        for color in colors {
            normalize_color(color);
        }
    }
}
